### PR TITLE
Show messages for users if the ROOT_URL is wrong, show JavaScript errors

### PR DESCRIPTION
**Pull request by wxiaoguang**
# ROOT_URL issues

Many maintainers have answered the issues about incorrect ROOT_URL again and again.

Using `HTMLURL`(absolute URL) in templates is the root case that Gitea can only serve web under ROOT_URL. That's why we have to force users to set ROOT_URL correctly, and that's why this PR came. If all URLs in template can be refactored to use Relative URL correctly, then we do not need to force users set ROOT_URL correctly again and again.

Actually, at the moment, there are 2 major cases:
* The assets can not be loaded (`AppSubUrl != ""` and users try to access `http://host:3000/`)
* The `ROOT_URL` is wrong, then many URLs in Gitea are broken.

This PR can detect these problems, and show enough information to users. No i18n is needed because these messages should only be processed by Gitea admin.

# JavaScript error issues

And, there are many users affected by JavaScript errors:

* https://github.com/go-gitea/gitea/issues/19245
* https://github.com/go-gitea/gitea/issues/19205
* https://github.com/go-gitea/gitea/issues/18991

And more and more

If these problems (JS errors) can be found at first time, then maintainers do not need to ask about how bug occurs again and again.

# Screenshots:

### Can not load `index.js`

<details>

image

</details>

### Incorrect ROOT_URL

<details>

image

</details>

### JavaScript error 

mainly caused by incorrect customized templates, or incorrect cached files

<details>

image

</details>

### `window.config` broken

<details>

image

image

</details>

# FAQ

## What if some users do not want to see these errors?

There is a specialized CSS class "js-global-error" for the error message, then end users still have a chance to hide these error messages by customized CSS styles. But please DO NOT do so, otherwise if the users ignore the warning, there will be a lot of UI problem on their side, for example: the issue preview / inline PR comment don't work, etc.

## What if I see "Your ROOT_URL in app.ini is undefined"

There must be something wrong with your JS assets or templates. Check your CDN/static files/customized templates, etc. Make sure the JS assets and templates are correct. 

## What if I see "Your ROOT_URL in app.ini is https://..... but you are visiting https://....."

Check the ROOT_URL config in your `app.ini`

```
[server]
ROOT_URL = https://.....
```

And check Gitea's startup logs:

```
AppURL(ROOT_URL): https://......
```

Make sure the ROOT_URL matches the base URL you are visiting.

## What if I see "JavaScript error"

It could be caused by:
* (either) incorrect template/JS assets on your side (including CDN/cache problems)
* (or) a frontend bug in Gitea or in your customized JS code

### CODE REVIEW

File: web_src/js/features/common-global.js
Comment (Gusted) [inline web_src/js/features/common-global.js]: Should these be displayed to all users?

Comment (wxiaoguang) [reply web_src/js/features/common-global.js]: Why not?

Comment (Gusted) [reply web_src/js/features/common-global.js]: I don't think it's appropriate to annoy the user with this error when they can't fix it anyway.

Can this JS can refactored away and only use go, by using `ctx.Flash`

Comment (wxiaoguang) [reply web_src/js/features/common-global.js]: It doesn't annoy any user, if a user sees this notice, the web has been wrong and doesn't work correctly for them. It's good for them to ask admin for help. I do not have other ideas about further changes.

Comment (Gusted) [reply web_src/js/features/common-global.js]: If it's being displayed to all users, then the message should be changed according to it, currently it sounds like the user has done something wrong, also then it should be i18n then IMO.

Comment (wxiaoguang) [reply web_src/js/features/common-global.js]: Why a normal user can see this warning on a correct Gitea instance?

This message should only be shown for the first installation, Gitea admin must be the first one to read it. If the Gitea admin doesn't handle it, it is the admin's problem.

If admin has done everything correctly, there should be no such warning.

Comment (wxiaoguang) [reply web_src/js/features/common-global.js]: TBH I think I have got full thought about this problem and I do not want to make further changes if there is no obvious bug. 

The purpose of this PR is to help maintainers to save time and help users to setup Gitea correctly.

So if anyone wants to improve, just replace this PR or work on it directly (including the texts, because I am not a native English speaker ... 😂)

**Review by Gusted — APPROVED** [review]

**Comment by wxiaoguang** [comment]
@go-gitea/maintainers I think we need this PR to save all active maintainers' time.

**Comment by lunny** [comment]
How about to force redirect to `ROOT_URL` when visit from a non-`ROOT_URL` address?

**Review by delvh — APPROVED** [review]

**Comment by wxiaoguang** [comment]
> How about to force redirect to `ROOT_URL` when visit from a non-`ROOT_URL` address?

It will cause problems if the ROOT_URL is `localhost` but the user is visiting Gitea from another host.

**Comment by wxiaoguang** [comment]
And I think we can wait for a few days before merge after lgtm/done. 

Maybe some maintainers have new options. 🙏

File: templates/base/head_script.tmpl
Comment (silverwind) [inline templates/base/head_script.tmpl:8]: Should probably create a separate webpack chunk that loads before`index.js` for this instead of inlining it. That way, you also shouldn't need this global variable.

Comment (wxiaoguang) [reply templates/base/head_script.tmpl:8]: However, `index.js` is very late, the layout is:

```
window.addEventListener('error')
window.config =  // error occurs 
template <script> // error occurs
index.js 
- bootstrap.js (the old publicpath.js)
- others (initXxx)  // error occurs
```

So the `window.addEventListener` must be the first one in script tag.

**Comment by Gusted** [comment]
> How about to force redirect to `ROOT_URL` when visit from a non-`ROOT_URL` address?

That will be putting up a bandaid around _possible_ incorrect configuration. To show the error seems the best option IMO. If they suddenly will be redirected to a new URL, how will they know what's going on and how to fix it?

**Comment by wxiaoguang** [comment]
Let's merge and see user's reactions. If it helps, that's good. If some users feel annoying, we should find the real problem behind these users.

**Comment by wxiaoguang** [comment]
FYI:

Using HTMLURL in templates is the root case that Gitea can only serve web under ROOT_URL. That's why we have to force users to set ROOT_URL correctly, and that's this PR came.

If all URLs in template can be refactored to use Relative URL correctly, then we do not need to force users set ROOT_URL correctly again and again.

**Comment by wxiaoguang** [comment]
Some updates about this PR (which may be useful for future readers).

The `show JavaScript errors` really helps to catch some frontend bugs, including:

* #19414
* #19494

And

* #19995
* #19996


The ROOT_URL detection also helps issues like #19972 and some users in discord.

----

I believe it's on the right way. If there are new  issues about it in the future, these related bugs could also be fixed.

**Pull request merged**
